builtin/logical/database/credentials.go

```diff
@@ -0,0 +1,172 @@
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/fatih/structs"
+	"github.com/hashicorp/vault/helper/random"
+	"github.com/mitchellh/mapstructure"
+)
+
+// passwordGenerator generates password credentials.
+// A zero value passwordGenerator is usable.
+type passwordGenerator struct {
+	// PasswordPolicy is the named password policy used to generate passwords.
+	// If empty (default), a random string of 20 characters will be generated.
+	PasswordPolicy string `mapstructure:"password_policy" structs:"password_policy,omitempty"`
+}
+
+// newPasswordGenerator returns a new passwordGenerator using the given config.
+// Default values will be set on the returned passwordGenerator if not provided
+// in the config.
+func newPasswordGenerator(config map[string]string) (passwordGenerator, error) {
+	var pg passwordGenerator
+	if err := mapstructure.WeakDecode(config, &pg); err != nil {
+		return pg, err
+	}
+
+	return pg, nil
+}
+
+// Generate generates a password credential using the configured password policy.
+// Returns the generated password or an error.
+func (pg passwordGenerator) generate(ctx context.Context, b *databaseBackend, wrapper databaseVersionWrapper) (string, error) {
+	if !wrapper.isV5() && !wrapper.isV4() {
+		return "", fmt.Errorf("no underlying database specified")
+	}
+
+	// The database plugin generates the password if its interface is v4
+	if wrapper.isV4() {
+		password, err := wrapper.v4.GenerateCredentials(ctx)
+		if err != nil {
+			return "", err
+		}
+		return password, nil
+	}
+
+	if pg.PasswordPolicy == "" {
+		return random.DefaultStringGenerator.Generate(ctx, b.GetRandomReader())
+	}
+	return b.System().GeneratePasswordFromPolicy(ctx, pg.PasswordPolicy)
+}
+
+// toKVPairs returns the configuration of the passwordGenerator
+// as a map from string to string.
+func (pg passwordGenerator) toKVPairs() map[string]string {
```

Comment:
Nit: `configMap` might be a bit more descriptive than `toKVPairs` on what this is returning.

Reply:
I like that better. Done in e9e3536.